test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,505 @@
+    local ok, err = pcall(box.ctl.promote)
+    luatest.assert(ok, string.format(
+        'error while promoting unconfigured server: %s', err))
+
+    local ok, err = pcall(box.ctl.demote)
+    luatest.assert(ok, string.format(
+        'error while demoting unconfigured server: %s', err))
+end
+
+-- Promoting current raft leader and synchro queue owner should succeed
+-- with elections enabled.
+g_nspaces.test_leader_promote = function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+    g.cluster:promote(g.server_1)
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while promoting leader with elections on: %s', err))
+
+    g.server_1:box_config({election_mode = 'off'})
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while promoting leader with elections off: %s', err))
+end
+
+g_nspaces.after_test('test_leader_promote', function(g)
+    g.cluster:demote(g.server_1)
+    g.server_2:wait_synchro_queue_owner_id(0)
+end)
+
+-- Demoting current follower should succeed.
+g_nspaces.test_follower_demote = function(g)
+    local ok, err = g.server_2:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while demoting follower with elections off: %s', err))
+
+    g.server_1:box_config({election_mode = 'manual'})
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while demoting follower with elections on: %s', err))
+end
+
+-- Promoting current raft leader should succeed,
+-- even if he doesn't own synchro queue with elections enabled.
+g_nspaces.test_raft_leader_promote = function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+
+    -- Promote server, but get stuck before obtaining synchro queue
+    -- (write term bump to wal, get stuck on promote)
+    wal_delay_start(g.server_1, 2)
+    g.server_1:exec(function()
+        box.ctl.promote()
+    end)
+    g.server_1:wait_wal_delay()
+    g.server_1:wait_election_state('leader')
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+
+    luatest.assert(ok, string.format(
+        'error while promoting raft leader: %s', err))
+end
+
+g_nspaces.after_test('test_raft_leader_promote', function(g)
+    -- Finish promoting server_1
+    wal_delay_end(g.server_1)
+    g.server_1:wait_synchro_queue_owner_id()
+
+    g.server_1:box_config({election_mode = 'off'})
+    g.cluster:demote(g.server_1)
+    g.server_2:wait_synchro_queue_owner_id(0)
+end)
+
+-- Promoting and demoting should work when everything is ok.
+g_nspaces.test_ok = function(g)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
```

Comment:
Why didn't you use `g.server_1:promote()`? Same in other places where you do `exec` manually and without that retry loop which is present in `server:promote()`.
Same for demotion.